### PR TITLE
Add ONNX Model Export to New Frontend

### PR DESCRIPTION
**Pull request by rayankrish**
Add model export functionality with relevant helper functions and tests.

### CODE REVIEW

File: orttraining/orttraining/python/training/orttrainer.py [hunk 1]
```diff
@@ -101,17 +107,18 @@ class ORTTrainer(object):
                                                              { 'params' : ['model_param1' , 'model_param_2'], 'alpha' : 0.0}
                                                            ],
                                             alpha=0.9, beta=0.999)
-            ort_trainer = ORTTrainer(model, model_desc, optim_config)
+            ort_trainer = ORTTrainer(model, model_desc, optim_config, loss_fn)
     """
+    DEFAULT_OPSET_VERSION = 12
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/python/training/orttrainer.py:112]: we are not using this global anymore. you can find opset on `ORTTrainer.options._internal_use.opset_version`

File: orttraining/orttraining/python/training/orttrainer.py [hunk 2]
```diff
@@ -140,7 +147,13 @@ def __init__(self, model, model_desc, optim_config, loss_fn=None, options=None):
 
         self.model_desc = _ORTTrainerModelDesc(model_desc)
         self.optim_config = optim_config
-        self.options = ORTTrainerOptions(options)
+        
+        if options:
+            self.options = ORTTrainerOptions(options)
+        else:
+            self.options = ORTTrainerOptions()
+        self._enable_internal_postprocess=True
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/python/training/orttrainer.py:155]: Not needed: It is on `ORTTrainer.options._internal_use.enable_internal_postprocess`

File: orttraining/orttraining/python/training/orttrainer.py [hunk 3]
```diff
@@ -140,7 +147,13 @@ def __init__(self, model, model_desc, optim_config, loss_fn=None, options=None):
...
+        self._extra_postprocess=None
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/python/training/orttrainer.py:156]: Not needed: It is on ORTTrainer.options._internal_use.extra_postprocess

File: orttraining/orttraining/python/training/orttrainer.py [hunk 4]
```diff
@@ -165,6 +178,206 @@ def save_as_onnx(self, path):
...
         pass
 
+    def _combine_torch_model_with_loss(self, model, loss_fn, input_names): #rewrite to grab these from self
+        # don't need to wrap anything
+        if not loss_fn:
+            return model
+
+        sig = signature(model.forward)
+        ordered_list_keys = list(sig.parameters.keys())
+        if loss_fn:
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/python/training/orttrainer.py:188]: Not needed because of the previous `if not loss_fn: return model`

File: orttraining/orttraining/python/training/orttrainer.py [hunk 5]
```diff
@@ -165,6 +178,206 @@ def save_as_onnx(self, path):
         """
         pass
 
+    def _combine_torch_model_with_loss(self, model, loss_fn, input_names): #rewrite to grab these from self
+        # don't need to wrap anything
+        if not loss_fn:
+            return model
+
+        sig = signature(model.forward)
+        ordered_list_keys = list(sig.parameters.keys())
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/python/training/orttrainer.py:187]: Maybe a better name, like `ordered_input_list`

File: orttraining/orttraining/python/training/orttrainer.py [hunk 6]
```diff
@@ -165,6 +178,206 @@ def save_as_onnx(self, path):
...
+        if loss_fn:
+            sig_loss = signature(loss_fn)
+            if len(sig_loss.parameters) != 2:
+                raise RuntimeError("loss function should take two arguments - predict and label.")
+
+            # label shall be the second input to loss_fn.
+            ordered_list_keys = [*ordered_list_keys, list(sig_loss.parameters.keys())[1]]
+        
+        match = True
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/python/training/orttrainer.py:196]: # Add a short comment in why this match is done
`match_input_name = True`

File: orttraining/orttraining/python/training/orttrainer.py [hunk 7]
```diff
@@ -165,6 +178,206 @@ def save_as_onnx(self, path):
...
+            if i < len(self.model_desc.inputs):
+                self.model_desc.add_type_to_input_description(i, sample_input.dtype)
+        for i, sample_output in enumerate(sample_outputs):
+            if i < len(self.model_desc.outputs):
+                self.model_desc.add_type_to_output_description(i, sample_output.dtype)
+        model.train()
+
+        f = io.BytesIO()
+
+        # Other export options to use(this is for backward compatibility).
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/python/training/orttrainer.py:275]: We are not supporting pytorch 1.4, so you can remove from lines 275-284 and just add `training=torch.onnx.TrainingMode.TRAINING` into `torch.onnn._export()` call

File: orttraining/orttraining/python/training/orttrainer.py [hunk 8]
```diff
@@ -165,6 +178,206 @@ def save_as_onnx(self, path):
...
+                input = tuple(args[0])
+            else:
+                input = args
+            """
+            input = args #added
+
+            for input_desc in input_desc_with_:
+                if input_desc[0] in kwargs:
+                    input = input + (kwargs[input_desc[0]],)
+            if internal_learning_rate is not None:
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/python/training/orttrainer.py:345]: 345-348 belongs to another PR.

345: is for train_step
347: is for loss scaling

File: orttraining/orttraining/python/training/orttrainer.py [hunk 9]
```diff
@@ -165,6 +178,206 @@ def save_as_onnx(self, path):
...
+                # loss_scale input name is needed to call train_step, for example:
+                #   kwargs[model.loss_scale_input_name] = loss_scale
+                #   outputs = model.train_step(*args, **kwargs)
+                # However, when first time train_step is called model.loss_scale_input_name is not set.
+                # To workaround this problem, we use the special name 'default_loss_scale_input_name' to indicate
+                # the loss_scale.
+                if 'default_loss_scale_input_name' in kwargs.keys():
+                    input = input + (kwargs['default_loss_scale_input_name'],)
+            """
+            fetches = None
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/python/training/orttrainer.py:360]: fetches will be implemented/tested in a different pr too

File: orttraining/orttraining/python/training/orttrainer.py [hunk 10]
```diff
@@ -165,6 +178,206 @@ def save_as_onnx(self, path):
...
+
+            if self._enable_internal_postprocess:
+                self._onnx_model = postprocess.run_postprocess(self._onnx_model)
+
+            if self._extra_postprocess:
+                self._extra_postprocess(self._onnx_model)
+            return
+
+    # temp function for testing
+    def get_onnx(self):
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/python/training/orttrainer.py:378]: don  forget to remove it in the final version. we can always access it directly by accessing _oonx_model

File: orttraining/orttraining/test/python/pt_model.py [hunk 11]
```diff
@@ -0,0 +1,61 @@
+import math
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/test/python/pt_model.py:1]: let's move this file to <repo>/examples/python/pytorch_transformer

File: orttraining/orttraining/test/python/orttraining_test_orttrainer_frontend.py [hunk 12]
```diff
@@ -404,3 +443,93 @@ def testLRSchedulerUpdateImpl(lr_scheduler, expected_values):
         assert len(lr_list) == 1
         assert_allclose(lr_list[0],
                         expected_values[step], rtol=rtol, err_msg="lr mismatch")
+
+def my_loss(x, target):
```
Comment (thiagocrepaldi) [inline orttraining/orttraining/test/python/orttraining_test_orttrainer_frontend.py:447]: my_loss, transformer_model_description, batchify, get_batch should be in a separate file in  /examples/python/pytorch_transformer
